### PR TITLE
refactor(loop): pseudocode-style run() body (#20)

### LINKED ISSUE #20: Refactor the main run() loop for readability (pseudocode-style body)
## Motivation

The main `run()` loop in [src/sabi/algorithms/loop.py](src/sabi/algorithms/loop.py) has accumulated significant complexity as v1 has filled in: tempering schedules, look-ahead acquisition states, two emulator paths per round (acquisition-state and round-end), cheap-update planning via `_plan_round_update`, dispatcher fallbacks to refit, metric protocol checks, and a separate final-evaluation pass at the un-tempered form.

Reading the body today, several round-internal phases are interleaved without clear boundaries:

1. State resolution (current state, look-ahead target state, invariance check).
2. Acquisition-state emulator + surrogate distribution materialization.
3. Acquisition itself.
4. Target evaluation on the new batch.
5. Round-end emulator update at `current_state`.
6. Round-end surrogate-distribution + estimate + metric evaluation.
7. Bookkeeping (per-round metrics row, tempering history).

The loop is correct, the helpers (`_build_surrogate_distribution`, `_plan_round_update`, `_evaluate_metrics`, `_check_protocols`) are well-factored, and the docstrings are excellent. But the **shape** of the per-round body is harder to scan than it should be — too much branching on `invariance.target_map` and tempering details mixed in with the algorithmic skeleton. A new reader (or the user themselves three months later) has to mentally unfold the round to recover "what does this algorithm actually do."

## Goal

Refactor the loop so the per-round body reads as high-level pseudocode — each line a named operation that corresponds to a step in a paper-style description of the algorithm. Push the conditional / dispatch / state-juggling complexity into named helpers whose docstrings explain what they do.

The target is something like:

```python
for round_idx in range(1, algorithm.n_rounds):
    round_state = _resolve_round_state(algorithm, round_idx)
    acq_emulator, Y_train_for_acq = _materialize_acquisition_view(
        emulator, emulator_state, round_state, X, Y_raw, ...
    )
    x_new, y_new_raw = _run_acquisition(algorithm, problem, acq_emulator, ..., key_acq)
    X, Y_raw, Y_train = _append_round_evaluations(X, Y_raw, x_new, y_new_raw, round_state)
    emulator, emulator_state = _update_round_end_emulator(
        emulator, emulator_state, round_state, X, Y_train, ...
    )
    estimate = algorithm.estimator(_build_round_end_posterior(...))
    per_round_metrics.append(_record_round_metrics(estimate, round_state, ..., key_metric))
    tempering_states.append(round_state.current)
```

The exact decomposition is a design call — the goal is "loop body reads like an outline of the algorithm," not "match this specific function list."

## What this is NOT

- **Not** a behavior change. After this refactor, every test passes unchanged, every smoke run produces bit-identical output (or near-identical, only differing on JAX recompilation noise that's documented). The shape of `RunResult` doesn't move.
- **Not** an opportunity to redesign the algorithm. State resolution rules (`AcquisitionTarget.CURRENT|NEXT|TERMINAL`), invariance optimizations, the two-path emulator handling, the un-tempered final evaluation — all preserved.
- **Not** a place to land new features. Checkpointing, cluster execution, and multi-output emulators each have their own issue. This refactor sets up cleaner hook points for those, but doesn't add them.

## Concrete deliverables

- `run()` body fits comfortably on one screen (target ~30-50 lines), and the per-round body is roughly the pseudocode shape above.
- The conditional reuse logic around `invariance` lives in a small named helper, not inline in the loop.
- The two emulator paths (acquisition-view and round-end) each have a single dedicated helper that owns the `_plan_round_update` + `update_emulator` dance.
- All existing tests pass without modification.
- A docstring on `run()` (or a `docs/loop.md` companion) walks through the round shape line-by-line so a new reader can map "loop code" ↔ "paper algorithm."

## Open design questions

- Should the per-round local state be promoted to a small `RoundState` dataclass (current/target tempering states, intermediates, invariance, current_state cached)? Probably yes — passing six positional args to each helper is itself a readability hit.
- Where does the look-ahead vs. round-end split actually live? Today it's two implicit "passes" in one for-body; making it two named functions is the cleanest framing but might introduce arg-passing bloat. A `RoundState` dataclass mitigates that.
- Does the tempering-related state belong on `RoundState`, or does it stay on `Algorithm` and get queried each step? Probably on `RoundState` — the loop re-queries the same `intermediate_target` calls from a few places today, and caching them in a struct is both faster and clearer.

## Out of scope

- Touching `Algorithm` or `RunResult` shape.
- Any changes to `_handlers`, `_plan_round_update` semantics, or the dispatcher in [src/sabi/emulators/dispatch.py](src/sabi/emulators/dispatch.py). The refactor is around the loop body, not under it.
- Extracting things into a class-based loop (`AlgorithmRunner` etc.). The current functional shape is fine; the issue is naming and structure within it.

### PR DESCRIPTION
**Pull request by arob5**
## Summary

Closes #20.

Refactor the per-round body of `run()` in `src/sabi/algorithms/loop.py` so it reads as a high-level outline of the algorithm, with one named helper per phase. Each helper's docstring explains what it does; the conditional / dispatch / state-juggling that was inline now lives in those helpers.

- Add a frozen `RoundState` dataclass caching `current_state`, `target_state`, both `IntermediateTarget`s, the per-axis `InvarianceFlags`, and `round_idx`.
- Extract the per-round phases into named helpers: `_resolve_round_state`, `_resolve_acquisition_view`, `_run_acquisition`, `_append_round_evaluations`, `_update_round_end_emulator`, `_build_round_metrics_row`.
- Factor round 0 (initial design) and the post-loop un-tempered final eval into `_run_initial_round` and `_run_final_eval` for symmetry.
- Update `run()`'s docstring with a paper-style outline mapping each phase to the helper that owns it.
- **Untouched:** `_check_protocols`, `_merge_no_overwrite`, `_evaluate_scheduled_metrics`, `_build_surrogate_distribution`, `_plan_round_update`, `_eval_round`. No changes to `Algorithm`, `RunResult`, the dispatcher, or `_handlers`.

Pure structural refactor — no behavior change, no API change, no test changes.

## Test plan

- [x] `scripts/python -m pytest tests/test_loop.py -q` — 6 passed
- [x] `scripts/python -m pytest tests/test_runner_build.py tests/test_scheduled_metrics.py tests/test_acquisition_target.py tests/test_dsp_gp.py -q` — 86 passed
- [x] `scripts/python -m pytest -q` (full suite) — 301 passed

### CODE REVIEW

**Comment by arob5** [comment]
# Review: PR #26 — refactor(loop): pseudocode-style run() body (#20)

**Summary.** Cleanly extracts the per-round body of `run()` into named helpers (`_resolve_round_state`, `_resolve_acquisition_view`, `_run_acquisition`, `_append_round_evaluations`, `_update_round_end_emulator`, `_build_round_metrics_row`) plus a `RoundState` value type, and lifts round 0 / final eval into `_run_initial_round` / `_run_final_eval`. Behavior preserved: PRNG split order, round-state resolution, two-path emulator handling, `tempering_states` ordering, `final_metrics` semantics, and the lazy SP/estimate construction all read 1:1 against the old code. The `run()` body now reads like the algorithmic skeleton — meets `contributing.md#main-loops-read-like-pseudocode`.

**CI status.** No checks configured on this repo; relying on author's local `pytest -q` (301 passed per PR body). Loop tests rerun locally: 6 passed.

## Blocking
None.

## Minor

- `src/sabi/algorithms/loop.py:232` — PRNG split is interleaved between phases (`_resolve_acquisition_view` → `jax.random.split` → `_run_acquisition`), which `contributing.md#main-loops-read-like-pseudocode` calls out specifically: "resolve keys at the top of the round or inside the helper that consumes them." Moving `key_acq, key_metric, key_loop = jax.random.split(key_loop, 3)` to the top of the for-body (right after `for round_idx in ...`) would fully comply and tighten the pseudocode read.

- `src/sabi/algorithms/loop.py:211` — Same point for round 0: the `key_metric_0, key_loop = jax.random.split(key_loop, 2)` happens in `run()` immediately before `_run_initial_round`, then both keys are passed in. Cleaner to split inside `_run_initial_round` (since it consumes both) and let `run()` pass only `key_loop`. Trade-off: that means `_run_initial_round` would have to thread the residual `key_loop` back out, which is uglier — fine to leave as-is, but worth noting it's the same interleaving pattern.

- `src/sabi/algorithms/loop.py:298` — `_run_initial_round` returns a 7-tuple including both `state_0` and `emulator_state` (which are the same value at this point). The docstring acknowledges the trade-off; consider returning a small `InitialRoundResult` dataclass if this grows. Non-blocking — the long unpacking at the call site is the only readability cost.

## Nits / suggestions

- `src/sabi/algorithms/loop.py:498` — Inconsistent positional vs keyword-only signatures across the new helpers: `_run_acquisition`, `_resolve_acquisition_view`, `_append_round_evaluations`, `_update_round_end_emulator` are positional; `_build_round_metrics_row`, `_run_final_eval`, `_eval_round` are keyword-only via `*`. Pick one (probably keyword-only for the larger ones — `_run_acquisition` already has 8 args).

- `src/sabi/algorithms/loop.py:666` — `_eval_round` still has untyped `target` and `current_intermediate` parameters. Since the PR now imports `TargetDistribution` and `IntermediateTarget` at the top, this is a free annotation upgrade. Pre-existing, untouched by the PR — fine to defer.

- `src/sabi/algorithms/loop.py:386` — `_run_final_eval`'s metric loop duplicates the protocol-check + context-build + apply-suffix + merge logic that `_evaluate_scheduled_metrics` does. The two differ only in that `_run_final_eval` hardcodes `metric_target=TERMINAL` and uses `final_estimate`/`final_sp` directly. Pre-existing duplication; the refactor would be a sensible place to consolidate (e.g., a helper that runs a fixed `(sp, estimate, metric_target)` against a metric list), but out of scope for a structural-refactor PR.

- The PR branch is on a stale merge-base (before address-issue / review skills landed on `main`). `gh pr diff` is clean (loop.py only), but local `git diff main...PR-branch` shows phantom adds for those files. Rebase-on-merge would clean it up; non-blocking since GitHub merges fine.

## What's good

- `RoundState` after the follow-up commit (`e83c131`) is a clean "one fact, one place" model — `current_intermediate.state` is the canonical state, no denormalized state fields.
- Docstrings on every new helper explain the *why* (invariance reuse, lazy SP construction, the un-tempered final eval's TERMINAL pin), not just the *what*. The `RoundState` docstring's "Attributes" block is a model for value types.
- The "Round shape (paper-style outline)" addition to `run()`'s docstring maps each pseudocode phase to its helper — exactly the documentation the contributing-doc invariant calls for.
- Behavior-preservation discipline: every conditional branch from the old code (`invariance.both`, `invariance.target_map`, `final_scheduled` empty, `tempering_states` empty fallback, the `RescaleOutputs(factor=1.0)` short-circuit) survives unchanged.

## Recommendation

**Approve with minor comments.** The PR cleanly delivers the issue-#20 deliverables. The PRNG-split-interleaving point is the only thing that's an explicit invariant violation per the docs; everything else is style or future-cleanup.
